Interactive filtering

arielsvn opened the pull request:
## Issue Number

close #780 
https://github.com/AlexsLemonade/refinebio-frontend/issues/374#issuecomment-436373470

## Purpose/Implementation Notes

This modifies the API to make the filters more interactive, when a filter in a category is applied the remaining categories will be updated. More details on https://github.com/AlexsLemonade/refinebio-frontend/issues/374#issuecomment-436373470

## Types of changes

- New feature (non-breaking change which adds functionality)
- Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Functional tests

Tested locally, although it would be good to see this in staging since I don't have much data.

## Checklist

- [ ] Lint and unit tests pass locally with my changes
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)
- [x] Any dependent changes have been merged and published in downstream modules

## Screenshots

![2018-11-13 12 47 36](https://user-images.githubusercontent.com/1882507/48432448-5a154000-e742-11e8-9570-981fa0d54fea.gif)

arielsvn commented:
Discussed this with @Miserlou: the current solution is relying on the order of the parameters in the URL to determine the last filter category that was applied. This isn't a good idea... will update the code to use a new parameter instead.

Miserlou commented on api/data_refinery_api/views.py:
```diff
@@ -152,21 +152,19 @@ class SearchAndFilter(generics.ListAPIView):
+    ordering_fields = ('total_samples_count', 'id', 'created_at',
+                       'source_first_published', 'accession_code',)
+    ordering = ('-total_samples_count',)
+
```
The ordering section is now repeated, so this code doesn't do anything

arielsvn replied:
good catch, not sure why this happened. I'll remove it

Miserlou commented on api/data_refinery_api/views.py:
```diff
@@ -152,21 +152,19 @@ class SearchAndFilter(generics.ListAPIView):
+
     # Ordering
     ordering_fields = ('total_samples_count', 'id', 'created_at', 'source_first_published', 'accession_code',)
-    samples_count = django_filters.NumberFilter(method='filter_samples_count')
```
Why was this removed without comment?

arielsvn replied:
I thought it wasn't needed, since we now use the field `total_samples_count`. Sorting by number of samples works locally for me.

I should have added a comment though.

arielsvn commented:
> Does this break the ability to filter/order by number of samples? 

It doesn't, at least locally I'm able to sort by the total number of processed samples. I'll double check.

> What does this do to performance?

Good question, it would be good to try it with more data. Although I was thinking, we can reduce the number of DB queries per group of filters, right now each group of filters is calculated twice.

arielsvn commented:
@Miserlou I addressed the comments and added a new parameter `filter_order` as we discussed. Interactive filtering will only be enabled when this parameter is provided.

In terms of performance still isn't great, but I reduced it to making a single db query per filter group as it was before.

arielsvn commented:
https://github.com/AlexsLemonade/refinebio-frontend/pull/432 updates the frontend to send the `filter_order` parameter introduced in this request.
